### DISCUSSION TITLE
kein Datenempfang von JK-B2A24S15P weder über UART noch über BLE. Befehle können an BMS gesendet werden

### DISCUSSION THREAD

**Discussion by Peet1985** [syssi/esphome-jk-bms#610] [Q&A]
Hallo,

ich empfange keine Daten vom BMS. Weder über UART noch über BLE. Ich habe es über einen ESP32 und über einen ESP8266 probiert. Nirgendwo ein Erfolg. Der Charge- oder Dischargestatus kann aber an das BMS gesendet werden.

Hat jemand eine Idee?

Log:
[logs_jk-bms_logs.txt](https://github.com/user-attachments/files/17344330/logs_jk-bms_logs.txt)

**Comment by syssi** [comment]
Lass es uns einmal gemeinsam per Bluetooth versuchen. Schnapp dir das BLE-Beispiel. Ändere die MAC-Adresse in den ersten Zeilen auf die deines BMS und poste hier ein Logbuch dieses Setups.

**Comment by Peet1985** [comment]
Das war eins meiner Probleme, wo bekomme ich die MAC denn her?

Ich meine mal gelesen zu haben, dass im DEBUG-Level eine Suche durchgeführt wird und das Gerät dann im LOG angezeigt wird. Bei mir kommt aber nichts.

**Reply by syssi** [reply]
Das ist korrekt. Wenn kein anderer BLE-Client mit deinem BMS verbunden ist, nur dann kündigt es sich als "Bereit für eine Verbindung" in der Luft an (BLE-Advertisement). Dieser Abschnitt ist dafür verantwortlich die in empfangenen Ankündigungen ins Logbuch zu schreiben:

```
esp32_ble_tracker:
  scan_parameters:
    active: false
  on_ble_advertise:
    then:
      - lambda: |-
          if (x.get_name().rfind("JK-", 0) == 0 || x.get_name().rfind("JK_", 0) == 0) {
            ESP_LOGI("ble_adv", "New JK-BMS found");
            ESP_LOGI("ble_adv", "  Name: %s", x.get_name().c_str());
            ESP_LOGI("ble_adv", "  MAC address: %s", x.address_str().c_str());
            ESP_LOGD("ble_adv", "  Advertised service UUIDs:");
            for (auto uuid : x.get_service_uuids()) {
              ESP_LOGD("ble_adv", "    - %s", uuid.to_string().c_str());
            }
          }
```

Was für ein ESP32-Modell besitzt du?

**Comment by Peet1985** [comment]
Ich nutze einen ESP32-C3

Ich habe gerade nochmal geschaut, es wird kein gerät ausgegeben.

[logs_jk-bms1_logs.txt](https://github.com/user-attachments/files/17345322/logs_jk-bms1_logs.txt)

**Reply by Peet1985** [reply]
Ich habe gerade gesehen, dass die HW-Version nicht passt. Da habe ich micht verschaut. Kann das daran liegen?

![JK-BMS](https://github.com/user-attachments/assets/811c3d4b-4e7a-4ee0-bea6-fe784645b656)

**Reply by syssi** [reply]
Nein, es wird am ESP32-C3 liegen. Ich rechne damit, dass es funktioniert, sobald du auf einen ESP32-WROOM wechselst.

**Reply by Peet1985** [reply]
Ich habe jetzt mal einen bestellt, der kommt morgen. Dann sehen wir weiter.

Danke schonmal

**Reply by Peet1985** [reply]
Der WROOM ist angekommen und das bild war das selbe, es wurd kein Gerät gefunden. Ich habe dann über die Android-APP "BLE Scanner" die MAC-Adresse raus bekommen und konnte eine Verbindung aufbauen (auch mit dem ESP32-C3). Allerdings ist die Verbindung nicht sehr stabil, ständig wird sich neu verbunden (man hört es piepen vom BMS), die Daten kommen sehr verzögert oder werden garnicht aktualisiert. Egal mit welchen ESP ich es teste, es ist bei beiden das selbe. 

Gibt es nicht eine Möglichkeit an die Daten über UART ran zu kommen?

**Reply by syssi** [reply]
Handelt es sich um ein Model mit CAN-Bus-Support oder ohne? Kannst du einen der ESP32 (nicht ESP8266!) an den UART-Port anschließen und zwar die GPIOs

```
tx_pin: GPIO16
rx_pin: GPIO17
````

und nicht `GPIO1` /  `GPIO3`.

**Reply by syssi** [reply]
Hintergrund: Der `ESP32-C3` ist ein Single-Core SoC. Der WROOM besitzt zwei Kerne und sollte deshalb stabiler laufen.

**Reply by Peet1985** [reply]
Das BMS hat auch CAN-Bus-Support aber ich gehe an UART über den GPS-Anschluss.

ich hatte bereits einen einen ESP32-C3 über UART angeschlossen aber das verhalten war das selbe wie mit dem ESP8266. Ich konnte Befehle ins BMS schicken (ich habe es über die APP gegengeprüft) aber ich habe keine Werte erhalten. Ich kann heute nachmittag das ganze gern nochmal mit einem WROOM prüfen aber ich denke das Ergebnis wird das selbe sein.

GPIO1 und 3 sind am C3 RX und TX.

**Reply by syssi** [reply]
Es gibt viele Nutzer, bei welchen der GPS-Port nicht funktioniert, wenn das BMS inklusive CAN-Support gekauft wurde. Deshalb die Fragestellung. Der Hardware-UART vom ESP32 verträgt sich besser mit dem BMS, deshalb der Vorschlag einen ESP32 zu verwenden anstatt einen ESP8266. Dann gibt es haeufig das Problem, dass an GPIO1/GPIO3 bereits der USB-to-TTL-Wandler haengt, welcher sich negativ auf den Betrieb auswirkt. Deshalb mein Vorschlag GPIOs zu nutzen, an welchen nicht bereits andere Komponenten hängen. Zu guter letzt muss man sich bei GPIO1/GPIO3 darum kümmern, dass der `logger` nicht ebenfalls auf diese Pins schreibt (was per Default der Fall ist).

TL&DR: Mit diesem Setup stehen die Chancen am besten, dass es klappt:

1. Ein BMS ohne CAN-Support (ignorieren wir in deinem Fall)
2. Einsatz eines ESP32 (WROOM immer ueber C3 bevorzugen). Der erste CPU-Kern kann sich ums WiFi kümmern. Auf dem zweiten Kern läuft die Anwendung / ESPHome. Muss sich ein Kern um Beides kümmern, muss man mit einbußen rechnen.
3. GPIO16 und GPIO17 nutzen

**Reply by Peet1985** [reply]
Habe es getestet wie beschrieben und Ergebnis ist das gleiche. Charge und Discharge kann geändert werden aber es werden keine Daten gelesen. Das heißt wohl, dass ich ein BMS habe, wo die Schnittstelle nicht richtig funktioniert?

**Reply by syssi** [reply]
Ich kann leider nur raten. Du bist sicher, dass die RX-Leitung am richtigen Pin der GPS-TTL-Buchse steckt?

**Reply by syssi** [reply]
Wenn due Fotos von deinem Aufbau bereitstellst, schaue ich gerne einmal drauf.

**Reply by Peet1985** [reply]
![20241014_173852](https://github.com/user-attachments/assets/8224e961-8840-4a0a-bdad-7e5458b2a775)

**Reply by Peet1985** [reply]
sollte passen.  RX und TX kann auch nicht vertauscht sein, sonst würde ich nichts zum BMS schicken können.

**Reply by syssi** [reply]
Sehe ich auch so. Du könntest das BMS abstecken und "gelb" + "grün" kurzschließen. Die periodischen Requests (auf TX) sollten daraufhin auf der RX-Leitung empfangen werden. Sie lassen sich zwar nicht decodieren aber wenn du den `debug`-Output der `uart`-Komponente aktiviert hast, dann solltest du den Traffic (`>>>` Outgoing traffic, `<<<` incoming traffic) im Logbuch sehen. Damit kannst du dich vergewissern, dass den ESP grundsätzlich in der Lage ist Daten zu empfangen.

**Reply by Peet1985** [reply]
Ich gehe mal davon aus, dass einer der drei ESPs die ich getestet habe, in der Lage war, Daten zu empfangen aber das BMS schickt einfach nichts raus. ich habe auch Kontakt zu Jikong aufgebaut und man sagte mir, dass man nicht garantieren kann, dass an irgendeinem BMS, UART am GPS-Port funktioniert. Da frage ich mich dann allerdings wie sichergestellt wird, dass der RS485 Adapter überall läuft.

Ich werde wohl auf BLE setzen müssen um das BMS in HA zu bekommen, wie frage ist, wie bekomme ich das ganze stabil?

**Reply by syssi** [reply]
Dann fangen wir doch nochmal mit BLE an. Bitte den ESP32-WROOM nutzen. Ein großes Log wäre super inkl. der Zeitstempel, wo das BMS einen Piep von sich gegeben hat. Dann versuchen wir in Erfahrung zu bringen, wo es klemmt.

Ich bin mit meinem Setup sehr zufrieden. Der ESP verliert ca. alle 3 Monate 1x die Verbindung zum BMS und berappelt sich nach wenigen Sekunden/Minuten. Das ist fuer mich voellig ausreichend.

**Reply by Peet1985** [reply]
Reicht es wenn der Akku im Standby ist oder muss er Arbeiten?

**Reply by syssi** [reply]
Die LED am BMS muss blinken. Dies zeigt die Bereitschaft zur Aufnahme einer Verbindung an. Leuchtet die LED dauerhaft, dann besteht eine aktive Verbindung. Das BMS kann genau eine Verbindung managen. Ist beispielsweise die Android App aktiv/eingebucht, so kann sich der ESP nicht mehr verbinden und andersrum.

**Reply by Peet1985** [reply]
Das weiß ich doch, ich meinte ob der Akku wärendessen ge-/entladen werden muss oder reicht es wenn er ohne last da steht.

**Reply by syssi** [reply]
Das ist egal. Das BMS überwacht permanent und ist währenddessen ansprechbar.

**Reply by Peet1985** [reply]
Sorry für die späte Antwort.

Ich schaffe es einfach nicht eine stabile Verbindung aufzubauen. Kaum verbindet er sich, reist die Verbindung schon wieder ab.
[logs_solarakku_logs(2).txt](https://github.com/user-attachments/files/17609792/logs_solarakku_logs.2.txt)
[logs_solarakku_logs(3).txt](https://github.com/user-attachments/files/17609793/logs_solarakku_logs.3.txt)

**Reply by syssi** [reply]
This is the important detail of the log:

```
[C8:47:8C:E2:8C:88] ESP_GATTC_DISCONNECT_EVT, reason 8
```

The meaning of `reason 8` is a connection timeout. What's the distance between the ESP and the BMS?

**Reply by Peet1985** [reply]
Ich habe schon mehrer Abstände getestet. Als der ESP direkt neben dem BMS lag war es ganz schlimm, jetzt bin ich ca 1,5m weg und es ist etwas besser geworden.

**Reply by Peet1985** [reply]
Bei so geringen Abständen sollte es doch nicht zu Problemen kommen oder? 

Wenn ich mich mit meinem Handy über die APP verbinde ist die Verbindung auch stabil. Somit glaube ich nicht, dass es am BMS liegt.

**Reply by syssi** [reply]
Am Ende ist es Funktechnik und man kann die beiden BLE-Chips der unterschiedlichen Geräte nicht so einfach vergleichen. Angenommen der ESP ist super empfindlich und das Signal vom BMS übersteuert bei kurzer Distanz so arg, dass es sich nicht mehr decodieren lässt, dann hilft Abstand.
